Let compaction wait on background deletions

bhalevy opened the pull request:
In several cases in distributed testing (dtest) we trigger compaction using nodetool compact assuming that when it is done, it is indeed really done.
However, the way compaction is currently implemented in scylla, it may leave behind some background tasks to delete the old sstables that were compacted.

This commit changes major compaction (triggered via the ss::force_keyspace_compaction api) so it would wait on the background deletes and will return only when they finish.
    
Fixes #4909

Tests: unit(dev), nodetool_refresh_with_data_perms_test, test_nodetool_snapshot_during_major_compaction

avikivity commented on table.cc:
```diff
@@ -1287,6 +1288,8 @@ table::compact_sstables(sstables::compaction_descriptor descriptor, bool cleanup
             }
+            // This is done in the background, so we can consider this compaction completed.
+            (void)this->delete_compacted_sstables(old_ssts);
         };
```
Maybe we shouldn't do it in the background, and consider all compactions complete after they deleted their inputs. If deletion is slow for some reason, and we don't wait for it, we can run out of space. Also background jobs are like land mines except they move around and actively seek you.

bhalevy replied:
I like that direction. It would also help with implementing a disk-space controller.

avikivity commented on sstables/compaction_manager.cc:
```diff
@@ -276,6 +276,7 @@ future<> compaction_manager::submit_major_compaction(column_family* cf) {
             // those are eligible for major compaction.
             sstables::compaction_strategy cs = cf->get_compaction_strategy();
             sstables::compaction_descriptor descriptor = cs.get_major_compaction_job(*cf, get_candidates(*cf));
+            descriptor.flags = sstables::compaction_flags::wait_for_deletion;
```
Can just use separate booleans, no need to descend to bit flipping.

bhalevy replied:
Let's go for the always synchronous solution and then no additional flags are needed.
Then, moving the bool cleanup parameter to `compaction_descriptor` could be done independently of this issue.

bhalevy commented:
> Can test_large_data in cql_query_test.cc be updated? Currently it has:
> 
> ```
>     // In order to guarantee that system.large_rows has been updated, we have to                                                                                                                                                          
>     // * flush, so that a thumbstone for the above delete is created.                                                                                                                                                                     
>     // * do a major compaction, so that the thumbstone is combined with the old entry.                                                                                                                                                    
>     // * stop the db, as only then do we wait for sstable deletions
> ```

@espindola sure, good idea. let's take that into account in the next version.

bhalevy commented on table.cc:
```diff
@@ -1267,29 +1268,48 @@ void table::remove_ancestors_needed_rewrite(std::unordered_set<uint64_t> ancesto
             this->on_compaction_completion(new_ssts, old_ssts);
             if (release_exhausted) {
                 release_exhausted(old_ssts);
             }
```
Ok. Given @avikivity's direction of always waiting for deletion we would probably solve it differently though.

bhalevy commented:
In v3:
- Made all compactions wait on deletes, not only major compaction
  (no optional/shared promise required)
- Dropped compaction_flags change.
- Dropped on_compaction_completion refactoring 
- Added "tests: test_large_data: do not stop database"

raphaelsc commented:
couldn't unconditionally waiting for deletion to complete become a problem with incremental compaction? Compaction would be blocked whenever it waits on deletion of a sstable which is held by any operation. For example, a time-consuming  read operation which holds a sstable to be released could block compaction for a considerable amount of time. So we could end up with a performance issue.

raphaelsc commented:
turns out sstable deletion doesn't rely on destruction of sstable object, so I think we can ignore my comment above

bhalevy commented:
> turns out sstable deletion doesn't rely on destruction of sstable object, so I think we can ignore my comment above

Right.  Otherwise, do you approve?  (If so, formally approving would be great :) )

bhalevy commented on table.cc:
```diff
@@ -1191,43 +1192,34 @@ table::on_compaction_completion(const std::vector<sstables::shared_sstable>& new
-            });
-            _sstables_compacted_but_not_deleted.erase(e, _sstables_compacted_but_not_deleted.end());
-            rebuild_statistics();
+    if (f.failed()) {
```
Right. We're not in `then_wrapped`...

bhalevy commented on tests/cql_query_test.cc:
```diff
@@ -121,16 +121,15 @@ SEASTAR_THREAD_TEST_CASE(test_large_data) {
                 return dbi.get_compaction_manager().submit_major_compaction(&*table.second);
             });
         }).get();
-        stop_database(e.db()).get();
```
Good idea.

bhalevy replied:
Opened #4936

avikivity replied:
sstables can be kept open by ongoing queries and by queriers.

avikivity replied:
@espindola I have a patchset that moves tracking of real sstable lifetime to sstables_manager, it can be used to improve this.

bhalevy commented:
In v4:
* on_compaction_completion: try { f.get(); } catch (...) rather than calling f.failed()
  - assert(thread::running_in_thread())
* cql_query_test: test_large_data: fix typo in comment: s/tobstone/tombstone/

espindola commented on table.cc:
```diff
@@ -1199,16 +1199,18 @@ table::on_compaction_completion(const std::vector<sstables::shared_sstable>& new
+    try {
+        f.get();
+    } catch (...) {
+        tlogger.error("Compacted SSTables deletion failed: {}. Ignored.", std::current_exception());
```
Should we always ignore this? Please add a comment as to why or a FIXME.

avikivity replied:
There's really no way to recover.

bhalevy replied:
I'll add a comment explaining this.

avikivity commented on table.cc:
```diff
@@ -1199,16 +1199,18 @@ table::on_compaction_completion(const std::vector<sstables::shared_sstable>& new
+           return sstables::delete_atomically(std::move(sstables_to_remove));
+       });
+    });
+
```
This was originally done in the background because the atomic_deletion_manager could impose unbounded latency, but now that it is gone, I don't think it's a problem.

avikivity commented on table.cc:
```diff
@@ -1220,14 +1222,6 @@ table::on_compaction_completion(const std::vector<sstables::shared_sstable>& new
             });
             _sstables_compacted_but_not_deleted.erase(e, _sstables_compacted_but_not_deleted.end());
             rebuild_statistics();
```
Does _sstables_compacted_but_not_deleted make any sense after this patch?

bhalevy replied:
@raphaelsc ^^^

bhalevy replied:
@avikivity
1. I think we still want operations like `sstable_info` to remain conservative and account for these sstables until they are deleted.
2. `get_fully_expired_sstables` algorithm takes `compacted_undeleted_sstables` into account so removing it entirely would be quite intrusive.

bhalevy commented on table.cc:
```diff
@@ -1199,16 +1199,18 @@ table::on_compaction_completion(const std::vector<sstables::shared_sstable>& new
+       });
+    });
+
+    assert(thread::running_in_thread());
```
OK, will revert assert and add comment back.

bhalevy commented:
In v5:
- removed assert(thread::running_in_thread())
- added comment about ignoring delete_atomically exception

slivne commented:
@avikivity ping, not clear to me if something is missing.

slivne commented:
@avikivity ping, not clear to me if something is missing.

avikivity commented:
> @avikivity ping, not clear to me if something is missing.

Nothing, thanks for raising.
